python/ray/data/_internal/datasource/delta_datasink.py

```diff
@@ -0,0 +1,3046 @@
+    file_options: Optional[Any] = None  # e.g. ParquetFileWriteOptions
+    max_partitions: Optional[int] = None
+    max_open_files: int = 1024
+    max_rows_per_file: int = 10 * 1024 * 1024
+    max_rows_per_group: int = 128 * 1024
+    min_rows_per_group: int = 64 * 1024
+
+    # --- Table (delta metadata) ---
+    name: Optional[str] = None
+    description: Optional[str] = None
+    configuration: Optional[Mapping[str, Optional[str]]] = None
+
+    # --- Schema/overwrite ---
+    overwrite_schema: bool = (
+        False  # When used, mapped to schema_mode='overwrite' in new API
+    )
+
+    # --- Storage/engine ---
+    storage_options: Optional[Dict[str, str]] = None
+    engine: Literal["pyarrow", "rust"] = "rust"
+    large_dtypes: bool = False
+    writer_properties: Optional[WriterProperties] = None  # Native Delta/WritersOptions
+    predicate: Optional[str] = None  # Overwrite predicate
+    target_file_size: Optional[int] = None
+
+    # --- Commit & transaction ---
+    commit_properties: Optional[CommitProperties] = None
+    post_commithook_properties: Optional[PostCommitHookProperties] = None
+
+    @classmethod
+    def from_dict(cls, config_dict: dict):
+        """
+        Create a DeltaWriteConfig from a dictionary. Only sets attributes present in the class.
+        """
+        obj = cls()
+        for key, val in config_dict.items():
+            if hasattr(obj, key):
+                setattr(obj, key, val)
+            else:
+                raise ValueError(f"Unknown config key: {key}")
+        return obj
+
+
+class DeltaUtilities:
+    def __init__(
+        self,
+        delta_uri: str,
+        schema: pa.schema,
+        filesystem: pa_fs.FileSystem,
+        partition_cols: Optional[List[str]] = None,
+        max_partitions: Optional[int] = None,
+        config: Optional[DeltaWriteConfig] = None,
+    ) -> None:
+        """
+        Handles writing data batches to a Delta Table
+
+        Examples:
+            .. testcode::
+                :skipif: True
+
+                writer = DeltaDatasetWriter(
+                    delta_uri="s3://bucket/path/to/delta-table",
+                    schema=schema,
+                    filesystem=fs,
+                    partition_cols=["country", "date"]
+                )
+                add_actions = writer.write(batch_ref, file_options=file_opts)
+
+        Args:
+            delta_uri: The root URI of the Delta table.
+            schema: The expected Arrow schema of the data.
+            filesystem: The Arrow-compatible filesystem object.
+            partition_cols: Columns to partition by; optional.
+            max_partitions: Upper bound for file parallelism (default 128).
+            config: Optional DeltaConfig object
+        """
+        self.delta_uri = delta_uri
+        self.schema = schema
+        self.filesystem = filesystem
+        self.partition_cols = partition_cols or []
+        self.max_partitions = max_partitions
```

Comment:
Just wondering: does this error out and abort the write if more partitions are discovered than the value set?